### PR TITLE
Add signing and verification to pubsub

### PR DESCRIPTION
**Pull request by NIC619**
Build on #360 .
- add `strict_signing: bool` attribute to `Pubsub`, default to `True`
    - if `strict_signing` is set to `True`, we sign every message we publish and verify signature of every message we receive
- implement `signature_validator`

### CODE REVIEW

**Comment by mhchia** [comment]
Would you mind having a rebase onto the master branch?

**Comment by NIC619** [comment]
@mhchia @ralexstokes Merged #360 and rebased. It should be ready for review.

File: tests/pubsub/test_pubsub.py [hunk 1]
```diff
@@ -510,3 +510,18 @@ def failed_sync_validator(peer_id, msg):
     await pubsubs_fsub[0].push_msg(pubsubs_fsub[0].my_id, msg_2)
     await asyncio.sleep(0.01)
     assert not event.is_set()
+
+
+@pytest.mark.parametrize("num_hosts, is_strict_signing", ((2, True),))
+@pytest.mark.asyncio
+async def test_strict_signing(pubsubs_fsub, hosts, monkeypatch):
```
Comment (mhchia) [inline tests/pubsub/test_pubsub.py]: Would you mind adding a case which publishes messages with a {no signature|invalid signature}?

File: libp2p/pubsub/pubsub.py [hunk 2]
```diff
@@ -505,18 +530,32 @@ def handle_subscription(
...
+                topicIDs=msg.topicIDs,
+                from_id=msg.from_id,
+                seqno=msg.seqno,
+            )
+            payload = (
+                PUBSUB_SIGNING_PREFIX.encode() + msg_without_key_sig.SerializeToString()
+            )
+            if not signature_validator(
+                deserialize_public_key(msg.key), payload, msg.signature
+            ):
```
Comment (mhchia) [inline libp2p/pubsub/pubsub.py]: Should we also verify `ID.from_pubkey(msg.from_id) == msg.key`?

Comment (NIC619) [reply libp2p/pubsub/pubsub.py]: Good call!

**Comment by NIC619** [comment]
I applied the PR feedback and also update the `interop` pubsub test.

File: libp2p/pubsub/pubsub.py [hunk 3]
```diff
@@ -505,18 +530,39 @@ def handle_subscription(
...
+            # First, construct the original payload that's signed by 'msg.key'
+            msg_without_key_sig = rpc_pb2.Message(
+                data=msg.data,
+                topicIDs=msg.topicIDs,
+                from_id=msg.from_id,
+                seqno=msg.seqno,
+            )
+            payload = (
+                PUBSUB_SIGNING_PREFIX.encode() + msg_without_key_sig.SerializeToString()
+            )
```
Comment (mhchia) [inline libp2p/pubsub/pubsub.py]: Would it be better to move these verification logics to `signature_validator` and change its function signature to `signature_validator(msg: rpc_pb2.Message)`? IMO it seems to be more instinct.

**Comment by carver** [comment]
I took care of it for this release, but just a reminder to add a blurb to `newsfragments/` that will be included with the release notes in future PRs. See some examples here: https://github.com/libp2p/py-libp2p/commit/23fa86979db43102e44c300b75008c6394c49b5f

and how they turned out:
https://py-libp2p.readthedocs.io/en/latest/release_notes.html#libp2p-v0-1-4-2019-12-12

**Comment by NIC619** [comment]
@carver Will do!

**Pull request merged**